### PR TITLE
Dockerize the development environment

### PR DESCRIPTION
**Pull request by hennevogel**

### CODE REVIEW

**Comment by hennevogel** [comment]
@openSUSE/open-build-service would be nice if you can try it and give me feedback

**Comment by lethliel** [comment]
fails with:

```
# sudo rake docker:up
docker build . -t openbuildservice/frontend --build-arg CONTAINER_USERID=0
Sending build context to Docker daemon  93.4 MB
Step 1 : FROM openbuildservice/base
Pulling repository docker.io/openbuildservice/base
Tag latest not found in repository docker.io/openbuildservice/base
rake aborted!
Command failed with status (1): [docker build . -t openbuildservice/fronten...]
/home/marco/git_r/open-build-service/Rakefile:6:in `block (2 levels) in <top (required)>'
Tasks: TOP => docker:up
(See full trace by running task with --trace)
```

**Comment by hennevogel** [comment]
@lethliel fixed.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,21 @@
+FROM openbuildservice/base
```
Comment (lethliel) [inline Dockerfile]: I think you will need a FROM openbuildservice/base:423 here.

Comment (hennevogel) [reply Dockerfile]: This will pull just latest...

Comment (lethliel) [reply Dockerfile]: I don't see it: https://hub.docker.com/r/openbuildservice/base/tags/ 
There is only the 423 tag.

Comment (hennevogel) [reply Dockerfile]: Man docker tags are hard to understand. They are there now sorry...

**Comment by lethliel** [comment]
now it works better. But a new error occured: 

```
bash contrib/bootstrap.sh
Setting up the database...
WARNING: You already have the config file database.yml, make sure it works with vagrant
Starting openbuildservice_db_1

ERROR: for db  Cannot start service db: driver failed programming external connectivity on endpoint openbuildservice_db_1 (f4b22c1b76cee3bb2d25289422ce007f57d9a96c7b3246d7d6f4b45a63fa5f90): Error starting userland proxy: listen tcp 0.0.0.0:3306: bind: address already in use
ERROR: Encountered errors while bringing up the project.
Starting openbuildservice_db_1

ERROR: for db  Cannot start service db: driver failed programming external connectivity on endpoint openbuildservice_db_1 (ec61352366a112f946c86f29c6ca0c99d003f2f6bbfaec04540bf44b7e354efe): Error starting userland proxy: listen tcp 0.0.0.0:3306: bind: address already in use
ERROR: Encountered errors while bringing up the project.
Setting up the rails app...
WARNING: You already have the config file options.yml, make sure it works with vagrant
Starting openbuildservice_db_1

ERROR: for db  Cannot start service db: driver failed programming external connectivity on endpoint openbuildservice_db_1 (58289783656be920cad506557caa0c642997bc484de7e4dac8c8759034a12558): Error starting userland proxy: listen tcp 0.0.0.0:3306: bind: address already in use
ERROR: Encountered errors while bringing up the project.
rake aborted!
Command failed with status (1): [bash contrib/bootstrap.sh...]
```

**Comment by hennevogel** [comment]
@lethliel you have a backend running on those ports? :-)

**Comment by lethliel** [comment]
On 09/05/2017 11:15 AM, Henne Vogelsang wrote:
> @lethliel <https://github.com/lethliel> you have a backend running on those ports? :-)

No. I had a local mariadb instance running. That was the problem. After stopping it i got one step further.
Should that really break rake docker:up?

My next problem: 

Setting up the database...
WARNING: You already have the config file database.yml, make sure it works with vagrant
warning: parser/current is loading parser/ruby24, which recognizes
warning: 2.4.0-compliant syntax, but you are running 2.4.1.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
rake aborted!
Mysql2::Error: Can't connect to local MySQL server through socket '/var/run/mysql/mysql.sock' (2 "No such file or directory")

Tasks: TOP => db:version
(See full trace by running task with --trace)

**Comment by hennevogel** [comment]
@lethliel well we are exposing all ports also locally, so yes. Looks like your database.yml is missing the 'host: db' setting.

**Comment by lethliel** [comment]
@hennevogel: 

ok. Perhaps we should document it somewhere. Also that docker-compose is needed.

I have the development environment running, but:

- `host: db` in database.yml was missing. (Should that go into git?)
- `source_host = backend` in options.yml was missing. (git?)
- the $ipaccess problem (as discussed yesterday)

**Comment by hennevogel** [comment]
@lethliel all 3 config changes are on my TODO for today :-) And docker-compose is documented in CONTRIBUTING.md

**Comment by lethliel** [comment]
@hennevogel: ups you are right. Did not read that.

**Comment by hennevogel** [comment]
@lethliel config changes are now also included

**Comment by hennevogel** [comment]
I've just discovered that building is not really possible inside a container. Will investigate

**Comment by lethliel** [comment]
I get a completely new error now: 

```
Using bundler 1.13.6
Errno::EACCES: Permission denied @ rb_sysopen -
/usr/lib64/ruby/gems/2.4.0/cache/concurrent-ruby-1.0.5.gem
An error occurred while installing concurrent-ruby (1.0.5), and Bundler cannot
continue.
Make sure that `gem install concurrent-ruby -v '1.0.5'` succeeds before
bundling.
The command '/bin/sh -c bundle config build.nokogiri --use-system-libraries; bundle install' returned a non-zero code: 5
rake aborted!
Command failed with status (5): [docker build . -t openbuildservice/fronten...]
```

**Comment by hennevogel** [comment]
@lethliel are you running as root again?

**Comment by lethliel** [comment]
@hennevogel: no. As user.

**Comment by mschnitzer** [comment]
Is it possible to run a single test? If so, how? I think we discussed that on Thursday. If it isn't possible yet, see it as a reminder ;-)

**Comment by evanrolfe** [comment]
@hennevogel `rake docker:up` has been running for ~3 hours so I'm assuming its stuck. This is the output:

```
[evan@obs2]$ rake docker:up
docker build . -t openbuildservice/frontend --build-arg CONTAINER_USERID=1000
Sending build context to Docker daemon  84.06MB
Step 1/12 : FROM openbuildservice/base
latest: Pulling from openbuildservice/base
0d47f1dd5765: Pull complete 
05213fc1e999: Pull complete 
21c4a5a5c10e: Pull complete 
c747e0871ed8: Pull complete 
ec56cc80185b: Pull complete 
Digest: sha256:00c5d0944b0a1e7b0f4e859f97358b1a11af18587d50135a7661d9cb42c88f1e
Status: Downloaded newer image for openbuildservice/base:latest
 ---> 1b5f2af625ea
Step 2/12 : ARG CONTAINER_USERID
 ---> Running in 8c01122c65cb
```

**Comment by eduardoj** [comment]
> `rake docker:up` has been running for ~3 hours so I'm assuming its stuck.

You've been bitten by the same problem as me. It happens when you use docker with btrfs. This is the bug where it is explained: https://bugzilla.suse.com/show_bug.cgi?id=1047152

I've solved this problem by creating an ext4 partition for `/var/lib/docker`.

**Comment by hennevogel** [comment]
Okay peepz, I think this is good to go now.

File: Rakefile [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  namespace :test do
+    desc 'Run our frontend tests in the docker container'
+    task :frontend do
+      sh "docker-compose exec frontend bundle exec rspec"
+    end
+    desc 'Run our backend tests in the docker container'
+    task :backend do
+      sh "docker-compose exec backend cd /obs; make -C src/backend test"
```
Comment (eduardoj) [inline Rakefile]: I think you meant:

`sh "docker-compose exec backend bash -c 'cd /obs; make -C src/backend test'"`

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -125,54 +125,41 @@ make sure you run all this from a filesystem that supports this.
...
     ```
 
 7. Check out your OBS frontend:
 You can access the frontend at [localhost:3000](http://localhost:3000). Whatever you change in your cloned repository will have effect in the development environment.
-**Note**: The vagrant instance is configured with a default user 'Admin' and password 'opensuse'.
+**Note**: The development environment is configured with a default user 'Admin' and password 'opensuse'.
 
 8. Building packages:
-     The easiest way to start building is to create an interconnect to build.opensuse.org. All resources, including the base distributions can be used that way directly.
+     The easiest way to start building is to create an interconnect to our refference server. All resources from the openSUSE instance, including the base distributions, can be used that way.
```
Comment (avindra) [inline CONTRIBUTING.md]: Typo (`reference`)

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -125,54 +125,41 @@ make sure you run all this from a filesystem that supports this.
    git submodule update
```
Comment (avindra) [inline CONTRIBUTING.md]: Is this step actually necessary anymore?

**Comment by evanrolfe** [comment]
@hennevogel the `rake docker:up` command fails on step 11 for me, this is the output below. I'm not sure how to debug this so let me know if you want me to run some commands or something..

```
Step 11/12 : RUN bundle config build.nokogiri --use-system-libraries; bundle install
 ---> Running in c5dc498705c3
Fetching gem metadata from https://rubygems.org/...........
Fetching version metadata from https://rubygems.org/...
Fetching dependency metadata from https://rubygems.org/..
Using rake 12.0.0
Using mysql2 0.4.9
Using bundler 1.13.6
Errno::EACCES: Permission denied @ rb_sysopen -
/usr/lib64/ruby/gems/2.4.0/cache/concurrent-ruby-1.0.5.gem
An error occurred while installing concurrent-ruby (1.0.5), and Bundler cannot
continue.
Make sure that `gem install concurrent-ruby -v '1.0.5'` succeeds before
bundling.
The command '/bin/sh -c bundle config build.nokogiri --use-system-libraries; bundle install' returned a non-zero code: 5
rake aborted!
Command failed with status (5): [docker build . -t openbuildservice/fronten...]
/home/evan/Code/obs2/Rakefile:6:in `block (2 levels) in <top (required)>'
/home/evan/.rvm/gems/ruby-2.4.0@global/gems/rake-12.0.0/exe/rake:27:in `<top (required)>'
Tasks: TOP => docker:build
(See full trace by running task with --trace)
```

**Comment by avindra** [comment]
I noticed intermittent permissions issues as well with the rake command, which I was able to resolve by `chown`ing a few files.

I wonder if using `rake` from inside the container, instead of the host machine would increase stability.

**Comment by hennevogel** [comment]
@avindra check the docu. this is on purpose :)

File: CONTRIBUTING.md (hunk 4 above)
Comment (hennevogel) [reply CONTRIBUTING.md]: sure, this is the build script

File: CONTRIBUTING.md (hunk 3 above)
Comment (hennevogel) [reply CONTRIBUTING.md]: fixed. thanks!

**Comment by hennevogel** [comment]
🔥 in the 🕳!

**Pull request merged**